### PR TITLE
Optimize JNI performance and fix Decimal validation

### PR DESCRIPTION
**Pull request by geekphilosophy**
Improve JNI call performance ~3x with reduced overhead. Add JMH benchmark suite for tracking JNI performance. Fix Decimal validator to accept leading zeros in corpus integration tests.

*Issue #, if available:*

*Description of changes:*
Cedar-Java JMH Benchmark Results — After Performance Fixes
============================================================
Date:    2026-04-15
Base:    1d141bc (main)
JDK:     21.0.10 (Corretto)
JMH:     1.37
OS:      macOS (Darwin 25.3.0, aarch64)
Config:  @Warmup(iterations=3, time=1) @Measurement(iterations=5, time=1) @Fork(1)

Changes applied:
  1. Remove thread::spawn per JNI call, use std::panic::catch_unwind instead (interface.rs)
  2. Remove redundant j_input_str.push(' ') (interface.rs)
  3. Eliminate double JSON deserialization (readTree+readValue → single readValue) (BasicAuthorizationEngine.java)
  4. Cache JNI version check in static initializer instead of per-call (BasicAuthorizationEngine.java)
  5. Single-pass escape detection in ValueDeserializer using node.size() + node.has() (ValueDeserializer.java)

Results (thorough run: 2 forks, 5 warmup, 20 measurement iterations):
Benchmark                                   Mode  Cnt    Score    Error  Units
AuthorizationBenchmark.isAuthorized_medium  avgt   40  122.606 ±  2.303  us/op
AuthorizationBenchmark.isAuthorized_small   avgt   40   25.416 ±  0.837  us/op
AuthorizationBenchmark.validate_small       avgt   40  159.865 ± 29.990  us/op

Comparison to baseline (1d141bc):
                          Baseline        After         Improvement
  isAuthorized_small      85.4 ± 3.4    25.4 ± 0.8     ~3.4x faster
  isAuthorized_medium    329.1 ± 258.4  122.6 ± 2.3    ~2.7x faster
  validate_small         353.3 ± 146.0  159.9 ± 30.0   ~2.2x faster

### CODE REVIEW

File: CedarJava/src/jmh/java/com/cedarpolicy/AuthorizationBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        policies.add(new Policy("permit(principal, action, resource);", "policy0"));
+        smallPolicySet = new PolicySet(policies);
+
+        smallEntities = new HashSet<>();
+        smallEntities.add(new Entity(principal, new HashMap<>(), new HashSet<>()));
+        smallEntities.add(new Entity(action, new HashMap<>(), new HashSet<>()));
+        smallEntities.add(new Entity(resource, new HashMap<>(), new HashSet<>()));
+    }
+
+    private void setUpMediumScenario() throws Exception {
```
Comment (mark-creamer-amazon) [inline CedarJava/src/jmh/java/com/cedarpolicy/AuthorizationBenchmark.java:107]: Nit: I think if we anticipate adding more scenarios, we might want to rename this (perhaps `basicPhotoFlashScenario`), but I think this is good as is

File: CedarJava/src/main/java/com/cedarpolicy/serializer/ValueDeserializer.java [hunk 2]
```diff
@@ -85,68 +85,47 @@ public Value deserialize(JsonParser parser, DeserializationContext context) thro
...
-                                    } else {
-                                        String msg = "Invalid Entity Type" + val.get("type").textValue();
-                                        throw new InvalidValueDeserializationException(parser, msg, node.asToken(), Map.class);
-                                    }
-                                }
+                // Single-pass: check for escape sequences via direct lookup, avoiding full iteration
+                if (node.size() == 1) {
+                    if (node.has(ENTITY_ESCAPE_SEQ)) {
+                        JsonNode val = node.get(ENTITY_ESCAPE_SEQ);
+                        if (val.isObject() && val.has("id") && val.has("type") && val.size() == 2) {
```
Comment (mark-creamer-amazon) [inline CedarJava/src/main/java/com/cedarpolicy/serializer/ValueDeserializer.java]: Nit: Looks like we can also shed the `val.isObject() &&` because [`val.has(...)`](https://www.javadoc.io/doc/com.fasterxml.jackson.core/jackson-databind/latest/com/fasterxml/jackson/databind/JsonNode.html#has-java.lang.String-) checks this already

> Method that allows checking whether this node is JSON Object node and contains value for specified property.

File: CedarJava/src/main/java/com/cedarpolicy/serializer/ValueDeserializer.java [hunk 3]
```diff
@@ -85,68 +85,47 @@ public Value deserialize(JsonParser parser, DeserializationContext context) thro
...
-                                    EntityIdentifier id = new EntityIdentifier(val.get("id").textValue());
-                                    Optional<EntityTypeName> type = EntityTypeName.parse(val.get("type").textValue());
-                                    if (type.isPresent()) {
-                                        return new EntityUID(type.get(), id);
-                                    } else {
-                                        String msg = "Invalid Entity Type" + val.get("type").textValue();
-                                        throw new InvalidValueDeserializationException(parser, msg, node.asToken(), Map.class);
-                                    }
-                                }
+                // Single-pass: check for escape sequences via direct lookup, avoiding full iteration
```
Comment (mark-creamer-amazon) [inline CedarJava/src/main/java/com/cedarpolicy/serializer/ValueDeserializer.java:82]: Nice refactor here! Deserializer changes look equivalent to me

File: CedarJava/src/main/java/com/cedarpolicy/serializer/ValueDeserializer.java
Comment (muditchaudhary) [inline CedarJava/src/main/java/com/cedarpolicy/serializer/ValueDeserializer.java]: Nit: We can get rid of `EscapeType` as it is not longer being used.

File: CedarJava/src/main/java/com/cedarpolicy/serializer/ValueDeserializer.java (hunk 3 above)
Comment (muditchaudhary) [reply CedarJava/src/main/java/com/cedarpolicy/serializer/ValueDeserializer.java:82]: +1

File: CedarJava/src/main/java/com/cedarpolicy/serializer/ValueDeserializer.java [hunk 4]
```diff
@@ -85,68 +85,47 @@ public Value deserialize(JsonParser parser, DeserializationContext context) thro
...
+                if (node.size() == 1) {
+                    if (node.has(ENTITY_ESCAPE_SEQ)) {
+                        JsonNode val = node.get(ENTITY_ESCAPE_SEQ);
+                        if (val.isObject() && val.has("id") && val.has("type") && val.size() == 2) {
+                            EntityIdentifier id = new EntityIdentifier(val.get("id").textValue());
+                            Optional<EntityTypeName> type = EntityTypeName.parse(val.get("type").textValue());
```
Comment (muditchaudhary) [inline CedarJava/src/main/java/com/cedarpolicy/serializer/ValueDeserializer.java:88]: Not required for this PR, but something I remembered while reviewing it: we can further optimize the deserializer by caching results from pure functions that cross the FFI boundary like `EntityTypeName.parse(_)` to avoid redundant and expensive FFI roundtrip costs for the same inputs.  Can probably extend this idea beyond deserializers.

File: CedarJava/src/main/java/com/cedarpolicy/BasicAuthorizationEngine.java [hunk 5]
```diff
@@ -42,14 +42,20 @@
...
 import com.fasterxml.jackson.core.JsonProcessingException;
-import com.fasterxml.jackson.databind.JsonNode;
 
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 
 /** An authorization engine that is compiled in process. Communicated with via JNI. */
 public final class BasicAuthorizationEngine implements AuthorizationEngine {
     static {
         LibraryLoader.loadLibrary();
+        String jniVersion = getCedarJNIVersion();
```
Comment (muditchaudhary) [inline CedarJava/src/main/java/com/cedarpolicy/BasicAuthorizationEngine.java:52]: Nice, this makes perfect sense. The per-call version check in `call` would only be necessary if the FFI library wasn't statically loaded. Since it is, checking once at initialization is sufficient.

File: CedarJava/src/main/java/com/cedarpolicy/value/Decimal.java [hunk 6]
```diff
@@ -36,9 +36,6 @@ public static boolean validDecimal(String d) {
                 return false;
             }
             d = d.trim();
-            if (d.length() > 21) {
```
Comment (muditchaudhary) [inline CedarJava/src/main/java/com/cedarpolicy/value/Decimal.java:39]: This check oversimplified the validation. 

> A `decimal` value can range from `-922337203685477.5808` to `922337203685477.5807.`
([source](https://docs.cedarpolicy.com/policies/syntax-datatypes.html#datatype-decimal))

But Cedar trims leading zeroes. So a value like `-000000922337203685477.5808` is perfectly valid (Checked on [Cedar policy playground](https://www.cedarpolicy.com/en/playground))  yet would fail `d.length() > 21`.

This went uncaught for ~2 years and only surfaced with the new corpus tests. Rather than just removing the check, we should fix it to correctly validate the range after stripping leading zeroes.

File: CedarJava/src/jmh/java/com/cedarpolicy/AuthorizationBenchmark.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+        EntityTypeName accountType = EntityTypeName.parse("Account").get();
+
+        EntityUID principal = userType.of("alice");
+        EntityUID action = actionType.of("View_Photo");
+        EntityUID resource = photoType.of("pic01");
+
+        mediumRequest = new AuthorizationRequest(principal, action, resource, new HashMap<>());
+
+        Set<Policy> policies = new HashSet<>();
+        policies.add(new Policy(
```
Comment (muditchaudhary) [inline CedarJava/src/jmh/java/com/cedarpolicy/AuthorizationBenchmark.java]: Nit: We can put the policies in a file and use `PolicySet.parsePolicies()` [Example](https://github.com/cedar-policy/cedar-java/blob/main/CedarJava/src/test/java/com/cedarpolicy/PolicySetTests.java#L40)

Same for Entities using `Entities.parse` [Example](https://github.com/cedar-policy/cedar-java/blob/main/CedarJava/src/test/java/com/cedarpolicy/EntitiesTests.java#L124)

File: CedarJavaFFI/src/interface.rs [hunk 8]
```diff
@@ -82,17 +78,14 @@ pub fn callCedarJNI(
...
+    let j_input_str: String = match env.get_string(&j_input) {
         Ok(s) => s.into(),
         Err(_) => return build_err_obj(&env, "parsing"),
     };
-    j_input_str.push(' ');
-
-    let handle = thread::spawn(move || call_cedar_in_thread(j_call_str, j_input_str));
 
-    let result = match handle.join() {
+    let result = match panic::catch_unwind(|| call_cedar(&j_call_str, &j_input_str)) {
```
Comment (muditchaudhary) [inline CedarJavaFFI/src/interface.rs:86]: This is nice.

Nit: `call_cedar_in_thread` is probably dead code now.

**Review by muditchaudhary — CHANGES_REQUESTED** [review]
Mostly nits except for decimal validation.

File: CedarJava/src/main/java/com/cedarpolicy/value/Decimal.java (hunk 6 above)
Comment (geekphilosophy) [reply CedarJava/src/main/java/com/cedarpolicy/value/Decimal.java:39]: Done, added specific tests for this.

File: CedarJava/src/test/java/com/cedarpolicy/DecimalTests.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+ */
+
+package com.cedarpolicy;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import org.junit.jupiter.api.Test;
+import com.cedarpolicy.value.Decimal;
+
+public class DecimalTests {
```
Comment (muditchaudhary) [inline CedarJava/src/test/java/com/cedarpolicy/DecimalTests.java:24]: We should also add tests for weird edge cases like `.`, `1.`, `-.0`, `1.2.3`, `-.`. Going through the decimal validator, I feel that some of these might not throw an exception and will require tightening up the regex.

File: CedarJava/src/main/java/com/cedarpolicy/value/Decimal.java [hunk 10]
```diff
@@ -30,21 +29,48 @@ public class Decimal extends Value {
...
+            String withoutSign = negative ? d.substring(1) : d;
+
+            int dotIndex = withoutSign.indexOf('.');
+            String intPart = withoutSign.substring(0, dotIndex);
+            String fracPart = withoutSign.substring(dotIndex + 1);
+
+            // Strip leading zeros from integer part
+            intPart = intPart.replaceFirst("^0+", "");
+            if (intPart.isEmpty()) {
+                intPart = "0";
```
Comment (muditchaudhary) [inline CedarJava/src/main/java/com/cedarpolicy/value/Decimal.java]: This will lead to accepting cases like `.1` which is invalid. Can tighten the upstream regex and not have this conditional instead.

File: CedarJava/src/main/java/com/cedarpolicy/value/Decimal.java [hunk 11]
```diff
@@ -30,21 +29,48 @@ public class Decimal extends Value {
...
+            // Pad fractional part to exactly 4 digits
+            fracPart = fracPart + "0000".substring(fracPart.length());
+
+            // Combine integer and fractional parts as the i64 representation (value * 10000)
+            String combined = intPart.equals("0") ? fracPart.replaceFirst("^0+", "") : intPart + fracPart;
+            if (combined.isEmpty()) {
+                combined = "0";
+            }
+
+            String limit = negative ? MAX_NEGATIVE : MAX_POSITIVE;
```
Comment (muditchaudhary) [inline CedarJava/src/main/java/com/cedarpolicy/value/Decimal.java]: Possibly range validation can be simplified using `BigDecimal`. 
```
BigDecimal val = new BigDecimal(d);
return val.compareTo(RANGE_MIN) >= 0 && val.compareTo(RANGE_MAX) <= 0;
```

Rest of the PR looks all good to me.

File: CedarJava/src/test/java/com/cedarpolicy/DecimalTests.java (hunk 9 above)
Comment (geekphilosophy) [reply CedarJava/src/test/java/com/cedarpolicy/DecimalTests.java:24]: Done, with a few more.

File: CedarJava/src/main/java/com/cedarpolicy/value/Decimal.java (hunk 11 above)
Comment (geekphilosophy) [reply CedarJava/src/main/java/com/cedarpolicy/value/Decimal.java]: Done, great simplification.

File: CedarJava/src/main/java/com/cedarpolicy/value/Decimal.java (hunk 10 above)
Comment (geekphilosophy) [reply CedarJava/src/main/java/com/cedarpolicy/value/Decimal.java]: Fixed.

**Review by muditchaudhary — APPROVED** [review]

**Review by mark-creamer-amazon — APPROVED** [review]
